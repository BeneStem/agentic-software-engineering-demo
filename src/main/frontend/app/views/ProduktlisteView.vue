<template>
  <div class="erk_finden__produktliste">
    <UeberschriftZwei
      class="erk_finden__produktliste-headline"
      :ist-haupt-ueberschrift="istHauptueberschrift"
      v-if="ueberschrift"
      >{{ ueberschrift }}
    </UeberschriftZwei>
    <div class="erk_finden__produktliste-produkte">
      <div
        class="erk_finden__produktliste-produkte-vogel"
        v-if="produkte && produkte.length > 2 && ueberschrift && !produktfilterAnzeigen"
      >
        <VogelMitHerzAnimiert />
      </div>
      <Produktfilter v-if="produktfilterAnzeigen" />
      <Produktliste :produkte="produkte" :trackingname="trackingname" :trackingnummer="trackingnummer" />
      <NulltrefferFilterMeldung v-if="zeigeNulltrefferseite" />
      <KeineProdukteMeldung v-if="!zeigeNulltrefferseite && produkte.length === 0" />
    </div>
    <div
      class="erk_finden__produktliste-alle-anzeigen-link-container"
      v-if="alleAnzeigenLinkText && alleAnzeigenLinkZiel"
    >
      <a
        class="erk_finden__produktliste-alle-anzeigen-link"
        :href="alleAnzeigenLinkZiel"
        @click="sendTrackAlleAnzeigenKlick"
        >{{ alleAnzeigenLinkText }}</a
      >
    </div>
  </div>
</template>

<script setup lang="ts">
import { computed, onServerPrefetch, watch } from 'vue';
import { useStore } from 'vuex';
import Produktliste from '@/main/frontend/app/components/Produktliste.vue';
import Produktfilter from '@/main/frontend/app/components/Produktfilter.vue';
import UeberschriftZwei from '@blume2000/design-system/dist/komponenten/content/UeberschriftZwei.vue';
import VogelMitHerzAnimiert from '@blume2000/design-system/dist/komponenten/atome/VogelMitHerzAnimiert.vue';
import NulltrefferFilterMeldung from '@/main/frontend/app/components/NulltrefferFilterMeldung.vue';
import KeineProdukteMeldung from '@/main/frontend/app/components/KeineProdukteMeldung.vue';
import { trackAlleAnzeigenKlick } from '../util/tracking';
import { FeatureToggles } from '@/main/frontend/app/util/featureToggles';
import { useProduktFilterTracking } from '@/main/frontend/app/composables/ProduktFilterTracking';
import { QueryParameters } from '@/main/frontend/app/model/QueryParameters';
import { useProduktFilter } from '@/main/frontend/app/composables/ProduktFilter';
import { useRoute } from 'vue-router';

const store = useStore();
const { sendNulltrefferTrackingEvent } = useProduktFilterTracking();
const { istFilterAktiv } = useProduktFilter();
const ueberschrift = computed(() => store.getters['EinstellungenState/getUeberschrift']);
const istHauptueberschrift = computed(() => store.getters['EinstellungenState/getIstHauptueberschrift']);
const trackingname = computed(() => store.getters['EinstellungenState/getTrackingname']);
const trackingnummer = computed(() => store.getters['EinstellungenState/getTrackingnummer']);
const produkte = computed(() => store.getters['ProdukteState/getProdukte']);
const alleAnzeigenLinkText = computed(() => store.getters['EinstellungenState/getAlleAnzeigenLinkText']);
const alleAnzeigenLinkZiel = computed(() => store.getters['EinstellungenState/getAlleAnzeigenLinkZiel']);
const produktfilterAnzeigen = computed<boolean>(
  () =>
    !store.getters['EinstellungenState/getIstFilterAusgeblendet'] &&
    store.getters['FeatureState/getFeatureStatus'](FeatureToggles.PRODUKTFILTER_ANZEIGEN)
);
const zeigeNulltrefferseite = computed(() => {
  return produktfilterAnzeigen.value && istFilterAktiv.value && produkte.value.length === 0;
});

watch(
  zeigeNulltrefferseite,
  (newValue, oldValue) => {
    if (newValue && !oldValue) {
      sendNulltrefferTrackingEvent();
    }
  },
  { immediate: true }
);

const sendTrackAlleAnzeigenKlick = trackAlleAnzeigenKlick;

onServerPrefetch(async () => {
  const route = useRoute();
  const queryParameters = QueryParameters.mapQueryToQueryParameters(route.query);

  // Diese Storeänderungen sind unabhängig voneinander und können daher gemeinsam awaited werden
  await Promise.all([
    store.commit('EinstellungenState/setEinstellungenFromQueryparameters', route.query),
    store.dispatch('FeatureState/fetchFeatureTogglesSSR'),
    store.commit('ProduktfilterState/setCmsFilter', queryParameters, { root: true }),
  ]);

  // Im folgenden sind wir abhängig von den vorangegangenen Änderungen und müssen daher sequentiel awaiten
  const produktfilterAnzeigen =
    !store.getters['EinstellungenState/getIstFilterAusgeblendet'] &&
    store.getters['FeatureState/getFeatureStatus'](FeatureToggles.PRODUKTFILTER_ANZEIGEN);

  if (produktfilterAnzeigen) {
    await store.dispatch('ProdukteState/fetchProdukteMitVerfuegbarenFilterwertenSSR', {
      istAktionAlleFilterZuruecksetzen: false,
      useDebouncedAPI: false,
    });
  } else {
    await store.dispatch('ProdukteState/fetchProdukteSSR');
  }
});
</script>

<style lang="scss">
.erk_finden__produktliste {
  margin: 0 auto;
  max-width: 100%;

  &-headline {
    margin-bottom: 24px;
  }

  &-produkte {
    position: relative;
    margin-top: 8px;
  }

  &-produkte-vogel {
    position: absolute;
    right: 12px;
    top: -66px;
    height: 66px;

    svg {
      transform: scale(-1, 1);
      --hauptfarbe: #{$color-dark};
    }
  }

  &-alle-anzeigen-link {
    @include font-Rubik-light(18);
    color: $color-dark;
    padding-left: 8px;
    padding-right: 8px;
    text-align: center;

    &:visited {
      color: $color-dark;
    }
  }

  &-alle-anzeigen-link-container {
    width: 100%;
    display: flex;
    justify-content: center;
    padding-top: 32px;
  }
}
</style>
