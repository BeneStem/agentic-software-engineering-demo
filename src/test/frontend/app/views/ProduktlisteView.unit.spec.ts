import { shallowMount } from '@vue/test-utils';
import ProduktlisteView from '@/main/frontend/app/views/ProduktlisteView.vue';
import { createStore } from 'vuex';

describe('ProduktlisteView.vue', () => {
  it('keine Ã¼berschrift wenn undefined', () => {
    // Given
    const store = createMockStore(false, false, false, true, false);
    const wrapper = shallowMount(ProduktlisteView, {
      global: {
        plugins: [store],
      },
    });

    // Then
    expect(wrapper.element).toMatchSnapshot();
  });

  it('zeigt ueberschrift wenn defined', () => {
    // Given
    const store = createMockStore(true, false, false, true, false);
    const wrapper = shallowMount(ProduktlisteView, {
      global: {
        plugins: [store],
      },
    });

    // Then
    expect(wrapper.element).toMatchSnapshot();
  });

  it('zeigt alle anzeigen link wenn text und ziel defined', () => {
    // Given
    const store = createMockStore(true, true, true, true, false);
    const wrapper = shallowMount(ProduktlisteView, {
      global: {
        plugins: [store],
      },
    });

    // Then
    expect(wrapper.element).toMatchSnapshot();
  });

  it('kein alle anzeigen link wenn ziel fehlt', () => {
    // Given
    const store = createMockStore(true, true, false, true, false);
    const wrapper = shallowMount(ProduktlisteView, {
      global: {
        plugins: [store],
      },
    });

    // Then
    expect(wrapper.element).toMatchSnapshot();
  });

  it('kein alle anzeigen link wenn text fehlt', () => {
    // Given
    const store = createMockStore(true, false, true, true, false);
    const wrapper = shallowMount(ProduktlisteView, {
      global: {
        plugins: [store],
      },
    });

    // Then
    expect(wrapper.element).toMatchSnapshot();
  });

  it('Sendet beim Klick auf den alle anzeigen Link ein Event an GA', () => {
    // Given
    const store = createMockStore(true, true, true, true, false);
    const wrapper = shallowMount(ProduktlisteView, {
      global: {
        plugins: [store],
      },
    });

    wrapper.vm.sendTrackAlleAnzeigenKlick = jest.fn();

    // When
    wrapper.find('a').trigger('click');

    // Then
    expect(wrapper.vm.sendTrackAlleAnzeigenKlick).toHaveBeenCalled();
  });

  it('Zeigt keine Filter an, wenn Produktfilter ausgeblendet ist und Feature Toggle profuktfilter_anzeigen aus', () => {
    // Given
    const store = createMockStore(true, false, false, true, false);
    const wrapper = shallowMount(ProduktlisteView, {
      global: {
        plugins: [store],
      },
    });

    // Then
    expect(wrapper.element).toMatchSnapshot();
  });

  it('Zeigt Filter an, wenn Produktfilter nicht ausgeblendet ist und Feature Toggle profuktfilter_anzeigen an', () => {
    // Given
    const store = createMockStore(true, false, false, false, true);
    const wrapper = shallowMount(ProduktlisteView, {
      global: {
        plugins: [store],
      },
    });

    // Then
    expect(wrapper.element).toMatchSnapshot();
  });

  it('Zeigt nulltreffer meldung wenn filter aktiv und dadurch keine Produkte sichtbar', () => {
    // Given
    const store = createMockStore(true, false, false, false, true, true);
    const wrapper = shallowMount(ProduktlisteView, {
      global: {
        plugins: [store],
      },
    });
    // Then
    expect(wrapper.element).toMatchSnapshot();
  });

  it('Zeigt keineprodukte meldung wenn keine filter aktiv und trotzdem keine Produkte sichtbar', () => {
    // Given
    const store = createMockStore(true, false, false, false, true, true);
    const wrapper = shallowMount(ProduktlisteView, {
      global: {
        plugins: [store],
      },
    });

    // Then
    expect(wrapper.element).toMatchSnapshot();
  });
});

function createMockStore(
  hasUeberschrift: boolean,
  hasLinkText: boolean,
  hasLinkZiel: boolean,
  istFilterAusgeblendet: boolean,
  isFeatureToggleActive: boolean,
  hatKeineProdukte = false,
  trackingname = 'trackingname',
  trackingnummer = 1234567890
) {
  const store = createStore({
    getters: {
      'EinstellungenState/getUeberschrift': (state: any): string | undefined => {
        return hasUeberschrift ? 'Ã¼berschrift' : undefined;
      },
      'ProdukteState/getProdukte': (state: any) => {
        if (hatKeineProdukte) return [];
        else {
          return [{}];
        }
      },
      'EinstellungenState/getAlleAnzeigenLinkText': (state: any): string | undefined => {
        return hasLinkText ? 'linktext' : undefined;
      },
      'EinstellungenState/getAlleAnzeigenLinkZiel': (state: any): string | undefined => {
        return hasLinkZiel ? 'linkziel' : undefined;
      },
      'EinstellungenState/getIstFilterAusgeblendet': (state: any): boolean => {
        return istFilterAusgeblendet ? true : false;
      },
      'FeatureState/getFeatureStatus':
        (state: any) =>
        (featureToggleName: string): boolean => {
          return isFeatureToggleActive;
        },
      'EinstellungenState/getTrackingname': (state: any): string => {
        return trackingname;
      },
      'EinstellungenState/getTrackingnummer': (state: any): number => {
        return trackingnummer;
      }
    },
  });
  store.dispatch = jest.fn();
  return store;
}
