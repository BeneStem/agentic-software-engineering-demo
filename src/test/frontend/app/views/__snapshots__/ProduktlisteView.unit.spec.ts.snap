// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProduktlisteView.vue Zeigt Filter an, wenn Produktfilter nicht ausgeblendet ist und Feature Toggle profuktfilter_anzeigen an 1`] = `
<div
  class="erk_finden__produktliste"
>
  <ueberschrift-zwei-stub
    class="erk_finden__produktliste-headline"
    isthauptueberschrift="false"
    istsemantischeueberschrift="true"
    mitrahmen="false"
    theme="hell"
  />
  <div
    class="erk_finden__produktliste-produkte"
  >
    <!--v-if-->
    <produktfilter-stub />
    <produktliste-stub
      produkte="[object Object]"
      trackingname="trackingname"
      trackingnummer="1234567890"
    />
    <!--v-if-->
    <!--v-if-->
  </div>
  <!--v-if-->
</div>
`;

exports[`ProduktlisteView.vue Zeigt keine Filter an, wenn Produktfilter ausgeblendet ist und Feature Toggle profuktfilter_anzeigen aus 1`] = `
<div
  class="erk_finden__produktliste"
>
  <ueberschrift-zwei-stub
    class="erk_finden__produktliste-headline"
    isthauptueberschrift="false"
    istsemantischeueberschrift="true"
    mitrahmen="false"
    theme="hell"
  />
  <div
    class="erk_finden__produktliste-produkte"
  >
    <!--v-if-->
    <!--v-if-->
    <produktliste-stub
      produkte="[object Object]"
      trackingname="trackingname"
      trackingnummer="1234567890"
    />
    <!--v-if-->
    <!--v-if-->
  </div>
  <!--v-if-->
</div>
`;

exports[`ProduktlisteView.vue Zeigt keineprodukte meldung wenn keine filter aktiv und trotzdem keine Produkte sichtbar 1`] = `
<div
  class="erk_finden__produktliste"
>
  <ueberschrift-zwei-stub
    class="erk_finden__produktliste-headline"
    isthauptueberschrift="false"
    istsemantischeueberschrift="true"
    mitrahmen="false"
    theme="hell"
  />
  <div
    class="erk_finden__produktliste-produkte"
  >
    <!--v-if-->
    <produktfilter-stub />
    <produktliste-stub
      produkte=""
      trackingname="trackingname"
      trackingnummer="1234567890"
    />
    <!--v-if-->
    <keine-produkte-meldung-stub />
  </div>
  <!--v-if-->
</div>
`;

exports[`ProduktlisteView.vue Zeigt nulltreffer meldung wenn filter aktiv und dadurch keine Produkte sichtbar 1`] = `
<div
  class="erk_finden__produktliste"
>
  <ueberschrift-zwei-stub
    class="erk_finden__produktliste-headline"
    isthauptueberschrift="false"
    istsemantischeueberschrift="true"
    mitrahmen="false"
    theme="hell"
  />
  <div
    class="erk_finden__produktliste-produkte"
  >
    <!--v-if-->
    <produktfilter-stub />
    <produktliste-stub
      produkte=""
      trackingname="trackingname"
      trackingnummer="1234567890"
    />
    <!--v-if-->
    <keine-produkte-meldung-stub />
  </div>
  <!--v-if-->
</div>
`;

exports[`ProduktlisteView.vue kein alle anzeigen link wenn text fehlt 1`] = `
<div
  class="erk_finden__produktliste"
>
  <ueberschrift-zwei-stub
    class="erk_finden__produktliste-headline"
    isthauptueberschrift="false"
    istsemantischeueberschrift="true"
    mitrahmen="false"
    theme="hell"
  />
  <div
    class="erk_finden__produktliste-produkte"
  >
    <!--v-if-->
    <!--v-if-->
    <produktliste-stub
      produkte="[object Object]"
      trackingname="trackingname"
      trackingnummer="1234567890"
    />
    <!--v-if-->
    <!--v-if-->
  </div>
  <!--v-if-->
</div>
`;

exports[`ProduktlisteView.vue kein alle anzeigen link wenn ziel fehlt 1`] = `
<div
  class="erk_finden__produktliste"
>
  <ueberschrift-zwei-stub
    class="erk_finden__produktliste-headline"
    isthauptueberschrift="false"
    istsemantischeueberschrift="true"
    mitrahmen="false"
    theme="hell"
  />
  <div
    class="erk_finden__produktliste-produkte"
  >
    <!--v-if-->
    <!--v-if-->
    <produktliste-stub
      produkte="[object Object]"
      trackingname="trackingname"
      trackingnummer="1234567890"
    />
    <!--v-if-->
    <!--v-if-->
  </div>
  <!--v-if-->
</div>
`;

exports[`ProduktlisteView.vue keine Ã¼berschrift wenn undefined 1`] = `
<div
  class="erk_finden__produktliste"
>
  <!--v-if-->
  <div
    class="erk_finden__produktliste-produkte"
  >
    <!--v-if-->
    <!--v-if-->
    <produktliste-stub
      produkte="[object Object]"
      trackingname="trackingname"
      trackingnummer="1234567890"
    />
    <!--v-if-->
    <!--v-if-->
  </div>
  <!--v-if-->
</div>
`;

exports[`ProduktlisteView.vue zeigt alle anzeigen link wenn text und ziel defined 1`] = `
<div
  class="erk_finden__produktliste"
>
  <ueberschrift-zwei-stub
    class="erk_finden__produktliste-headline"
    isthauptueberschrift="false"
    istsemantischeueberschrift="true"
    mitrahmen="false"
    theme="hell"
  />
  <div
    class="erk_finden__produktliste-produkte"
  >
    <!--v-if-->
    <!--v-if-->
    <produktliste-stub
      produkte="[object Object]"
      trackingname="trackingname"
      trackingnummer="1234567890"
    />
    <!--v-if-->
    <!--v-if-->
  </div>
  <div
    class="erk_finden__produktliste-alle-anzeigen-link-container"
  >
    <a
      class="erk_finden__produktliste-alle-anzeigen-link"
      href="linkziel"
    >
      linktext
    </a>
  </div>
</div>
`;

exports[`ProduktlisteView.vue zeigt ueberschrift wenn defined 1`] = `
<div
  class="erk_finden__produktliste"
>
  <ueberschrift-zwei-stub
    class="erk_finden__produktliste-headline"
    isthauptueberschrift="false"
    istsemantischeueberschrift="true"
    mitrahmen="false"
    theme="hell"
  />
  <div
    class="erk_finden__produktliste-produkte"
  >
    <!--v-if-->
    <!--v-if-->
    <produktliste-stub
      produkte="[object Object]"
      trackingname="trackingname"
      trackingnummer="1234567890"
    />
    <!--v-if-->
    <!--v-if-->
  </div>
  <!--v-if-->
</div>
`;
